Fix SST files not being cleaned up in the locations folder

danpi opened the pull request:
Descriptions of the changes in this PR:

<!-- Either this PR fixes an issue, -->

Fix #4554

### Motivation
Resolve issue with SST files remaining uncleared in the locations folder.


### Changes

1. Reuse the cleanup logic of `GarbageCollectorThread` to trigger entryLocationCompact() when a major compaction occurs.
2. Add the `entryLocationCompaction` configuration option to control the above behavior, disabled by default.

danpi commented:
I believe this issue is related to optimization #3653.

When using `deleteRange` to replace the one-by-one deletion method, only one action is written to RocksDB. If RocksDB is performing tiered compaction, this action is not guaranteed to be picked up, causing SST files to remain undeleted.

To resolve this issue:
Option 1: Revert to the previous logic of iteratively deleting each key.
Option 2: Use compactRange to forcefully clean up levels L0 to L6.
I chose to perform a compactRange synchronously when executing a major compaction.

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:
```diff
@@ -470,6 +473,10 @@ public void runWithFlags(boolean force, boolean suspendMajor, boolean suspendMin
                     gcStats.getMajorCompactionCounter().inc();
                     majorCompacting.set(false);
                 }
+                if (entryLocationCompaction) {
```
I suggest moving the entryLocationCompact out of the major or minor compaction scope and using a separate interval to control it.  Remove the `entryLocationCompaction` flag and use the entryLocationCompactionInterval instead. The default value is `-1`, which means disable. We can set it to 1 week by default in the future after verified this operation doesn't introduce a huge impact on performance.

The major or minor compaction has many impact factors, such as suspension, force, and interval.  It will make the entryLocation  compaction behavior unpredictable

danpi replied:
Great suggestion! Currently, the execution of `entryLocationCompact` is indeed coupled with the triggering of `majorCompaction`. I will modify it so that `entryLocationCompact` execution is only based on the configured trigger interval.

hangc0276 commented:
> I believe this issue is related to optimization #3653.
> 
> When using `deleteRange` to replace the one-by-one deletion method, only one action is written to RocksDB. If RocksDB is performing tiered compaction, this action is not guaranteed to be picked up, causing SST files to remain undeleted.
> 
> To resolve this issue: Option 1: Revert to the previous logic of iteratively deleting each key. Option 2: Use compactRange to forcefully clean up levels L0 to L6. I chose to perform a compactRange synchronously when executing a major compaction.

@danpi I think it should be a bug in RocksDB. We can use option 2

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:
```diff
@@ -470,6 +473,10 @@ public void runWithFlags(boolean force, boolean suspendMajor, boolean suspendMin
                 }
+                if (entryLocationCompaction) {
+                    // submit entryLocation compaction task
+                    ledgerStorage.entryLocationCompact();
```
Please add a log to show entryLocationCompact is triggered.

danpi commented:
@hangc0276 Addressed the feedback above. PTAL, thanks.

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:
```diff
@@ -87,6 +87,10 @@ public class GarbageCollectorThread implements Runnable {
     long majorCompactionMaxTimeMillis;
     long lastMajorCompactionTime;
 
+    boolean enableEntryLocationCompaction = false;
```
Remove this one and use `entryLocationCompactionInterval > 0` instead?

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:
```diff
@@ -489,6 +505,17 @@ public void runWithFlags(boolean force, boolean suspendMajor, boolean suspendMin
                     minorCompacting.set(false);
                 }
             }
+            if (enableEntryLocationCompaction && (curTime - lastEntryLocationCompactionTime
```
We'd better introduce a random factor for it. If we roll restart the BookKeeper cluster, and all the bookies start time are the same, which will make all the bookies triggered RocksDB compaction at the same time and impact the read latency.

hangc0276 replied:
The random factor can be any time in [0, entryLocationCompactionInterval]

hangc0276 commented:
> @hangc0276 Addressed the feedback above. PTAL, thanks.

@danpi Thanks for your contribution. I left more comments; please take a look. Thanks.

danpi commented:
> > @hangc0276 Addressed the feedback above. PTAL, thanks.
> 
> @danpi Thanks for your contribution. I left more comments; please take a look. Thanks.

@hangc0276 Add `randomCompactionDelay` to avoid all the bookies triggering compaction simultaneously. PTAL, thanks.

hangc0276 commented:
@danpi Would you please fix the check style issue?
```
Error:  /home/runner/work/bookkeeper/bookkeeper/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:220:30: '=' is not preceded with whitespace. [WhitespaceAround]
Error:  /home/runner/work/bookkeeper/bookkeeper/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:301: Line is longer than 120 characters (found 121). [LineLength]
Error:  /home/runner/work/bookkeeper/bookkeeper/bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java:3243: Line is longer than 120 characters (found 122). [LineLength]
Audit done.
[INFO] There are 3 errors reported by Checkstyle 9.3 with buildtools/src/main/resources/bookkeeper/checkstyle.xml ruleset.
Error:  src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:[220,30] (whitespace) WhitespaceAround: '=' is not preceded with whitespace.
Error:  src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:[301] (sizes) LineLength: Line is longer than 120 characters (found 121).
Error:  src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java:[3243] (sizes) LineLength: Line is longer than 120 characters (found 122).
```

horizonzy commented:
@danpi Could you help to fix the checkstyle.
